File: packages/watchmen-ai-copilot/src/watchmen_ai/hypothesis/model/hypothesis.py
from enum import Enum
from typing import Dict, List, Optional

from pydantic import BaseModel

from watchmen_ai.hypothesis.model.metrics import EmulativeAnalysisMethod
from watchmen_model.common import OptimisticLock, UserBasedTuple, Auditable
from watchmen_utilities import ExtendedBaseModel


class HypothesisStatus(str, Enum):
    DRAFTED = 'drafted'
    TESTING = 'testing'
    VALIDATED = 'validated'
    REJECTED = 'rejected'


class InsightType(str, Enum):
    RISK = 'risk'
    TRENDUP = 'trendup'
    OPPORTUNITY = 'opportunity'


class PriorityLevel(str, Enum):
    HIGH = 'high'
    MEDIUM = 'medium'
    LOW = 'low'


class RelatedHypothesis(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    status: HypothesisStatus
    confidence: float


class HypothesisContext(ExtendedBaseModel):
    source: str = 'manual'  # manual | chart | alert | chat
    sourceId: Optional[str] = None  # for chart source, the analysis (board) id
    metrics: Optional[List[str]] = None
    dimensions: Optional[List[str]] = None
    timeRange: Optional[str] = None
    filters: Optional[Dict[str, str]] = None


class Hypothesis(ExtendedBaseModel, UserBasedTuple, OptimisticLock,Auditable):
    id: Optional[str] = None
    title: str
    description: Optional[str] = None
    status: Optional[HypothesisStatus] = HypothesisStatus.DRAFTED
    confidence: Optional[float] = 0.0
    metrics: Optional[List[str]] = []
    businessChallengeId: Optional[str] = None
    relatedHypothesesIds: List[str] = []
    analysisMethod: Optional[EmulativeAnalysisMethod] = None
    context: Optional[HypothesisContext] = None




class Insight(BaseModel):
    id: str
    title: str
    type: Optional[InsightType] = None
    description: Optional[str] = None
    priority: Optional[PriorityLevel] = None

File: packages/watchmen-analysis-client/src/model/Hypothesis.ts
import { MetricDetail } from "./Metric";

export interface RelatedHypothesis {
    id: string;
    title: string;
    description: string;
    status: 'validated' | 'rejected' | 'testing' | 'drafted';
    confidence: number;
  }

export interface HypothesisContext {
    source: 'manual' | 'chart' | 'alert' | 'chat'; // where the hypothesis was proposed
    sourceId?: string; // analysisId / alertRuleId / chatMessageId
    metrics?: string[]; // metric names covered at proposal time (multiple for analysis-level hypotheses)
    dimensions?: string[];
    timeRange?: string;
    filters?: Record<string, string>;
}

export  interface HypothesisType {
    id: string;
    title: string;
    description: string;
    status: 'drafted' | 'testing' | 'validated' | 'rejected';
    confidence: number;
    metrics: string[];
    createdAt: string;
    businessChallengeId?: string; // Added business challenge reference
    relatedHypothesesIds?: string[]; // Added related hypotheses
    analysisMethod?: string; // Analysis method from EmulativeAnalysisMethod
    context?: HypothesisContext; // analysis scenario the hypothesis was proposed from
    // metrics_details?: MetricDetail[];
}


export interface HypothesisWithMetrics extends HypothesisType {
  metrics_details: MetricDetail[]
}


export interface Insight {
  title: string;
  type:"risk" | "trendup" | "opportunity";
  description: string;
  priority: 'high' | 'medium' | 'low';
}